src/aleph/services/p2p/protocol.py

```diff
@@ -29,100 +28,84 @@
-                                # remove all streams from that peer, ask to the others.
-                                for speer, info in list(streams):
-                                    if speer == peer:
-                                        streams.remove((speer, info))
-                                break
-
-                            return value
-                        except (StreamError):
-                            # let's delete this stream so it gets recreated next time
-                            # await stream.close()
-                            await stream.reset()
-                            streams.remove((peer, (stream, semaphore)))
-                            try:
-                                self.peers[peer].remove((stream, semaphore))
-                            except ValueError:
-                                pass
-                            LOGGER.debug("Can't request hash...")
-                await asyncio.sleep(0)
-
-            if not len(streams):
-                return
+        read_bytes = await stream.receive_some(MAX_READ_LEN)
+        if read_bytes is None:
+            return
+
+        try:
+            read_string = read_bytes.decode("utf-8")
+            message_json = json.loads(read_string)
+            if message_json["command"] == "hash_content":
+                value = await get_hash_content(
+                    message_json["hash"], use_network=False, timeout=1
+                )
+                if value is not None and value != -1:
+                    result = {
+                        "status": "success",
+                        "hash": message_json["hash"],
+                        "content": base64.encodebytes(value).decode("utf-8"),
+                    }
+                else:
+                    result = {"status": "success", "content": None}
+            elif message_json["command"] == "get_message":
+                result = {"status": "error", "reason": "not implemented"}
+            elif message_json["command"] == "publish_message":
+                result = {"status": "error", "reason": "not implemented"}
+            elif message_json["command"] == "hello":
+                result = {
+                    "status": "success",
+                    "content": {"version": __version__},
+                }
+            else:
+                result = {"status": "error", "reason": "unknown command"}
+            LOGGER.debug(f"received {read_string}")
+        except Exception as e:
+            result = {"status": "error", "reason": repr(e)}
+            LOGGER.exception("Error while reading data")
+
+        await stream.send_all(json.dumps(result).encode("utf-8"))
+
+    async def make_request(self, request_structure: Dict[str, Any]) -> Optional[Dict[str, Any]]:
+        peers = [peer for peer in self.peers]
+        # Randomize the list of peers to contact to distribute the load evenly
+        random.shuffle(peers)
+
+        for peer in peers:
```

Comment:
This was in a `while True`, why not anymore ?

Reply:
It was not a real while True, but more of a loop on all the streams opened with all the peers to see if someone has the content we are looking for. I just simplified it.